src/frontend/src/binder/expr/value.rs

```diff
@@ -110,6 +112,39 @@ impl Binder {
         Ok(expr)
     }
 
+    pub(super) fn bind_array_cast(&mut self, exprs: Vec<Expr>, ty: DataType) -> Result<ExprImpl> {
+        if exprs.is_empty() {
+            return Ok(FunctionCall::new_unchecked(
+                ExprType::Array,
+                vec![],
+                // Treat `array[]` as `varchar[]` temporarily before applying cast.
+                DataType::List {
+                    datatype: Box::new(ty),
+                },
+            )
+            .into());
+        }
+        let mut exprs = exprs
+            .into_iter()
+            .map(|e| self.bind_expr(e))
+            .collect::<Result<Vec<ExprImpl>>>()?;
+        // Cast to the same type.
+        for e in &mut exprs {
+            let mut dummy = ExprImpl::literal_bool(false);
+            swap(&mut dummy, e);
+            *e = dummy.cast_explicit(ty.clone())?;
+        }
```

Comment:
The propagation needs to happen recursively for multi-dimensional arrays. For example `select array[array[1, true]]::bool[][];`